File: source/solutions/aws-defaults/lambdas/aws-deeplens-image-capture-python/ggiot.py
import platform
import greengrasssdk
import json
from inspect import currentframe

class GGIoT:

    # Constructor
    def __init__(self, thing='default', prefix='mtm'):
        self.thing = thing
        self.prefix = prefix
        self.topicPrefix = self.prefix + '/' + self.thing + '/'
        self.topicLogger = self.topicPrefix + 'logger'

        if platform.system() != 'Darwin':
            GGC = greengrasssdk.client('iot-data')

            def prepPublish(topic=self.topicLogger, payload={}):
                GGC.publish(topic=topic, payload=json.dumps(payload))

            def prepUpdateShadow(thing=self.thing, payload={}):
                GGC.update_thing_shadow(thingName=thing, payload=json.dumps(payload))

            def prepGetShadow():
                response = GGC.get_thing_shadow(thingName=self.thing)
                payloadDict = json.loads(response['payload'])
                # stateDict = payloadDict['state']
                return payloadDict

            self.publish = prepPublish
            self.updateThingShadow = prepUpdateShadow
            self.getThingShadow = prepGetShadow

        else:
            print('Platform is Darwin (ie. could be a mac)')

            def debug(topic=self.topicLogger, payload={}):
                print(topic + ': ' + json.dumps(payload))

            def debugUpdateShadow(thing=self.thing, payload={}):
                print("updateThingShadow: " + thing + ": " + json.dumps(payload))

            def debugGetShadow(thing=self.thing, payload={}):
                print("getThingShadow: " + thing + ": {}")
                return {}

            self.publish = debug
            self.updateThingShadow = debugUpdateShadow
            self.getThingShadow = debugGetShadow

    def info(self, data):
        self.publish(topic=self.topicLogger, payload={
            "type":  "info",
            "payload": data
        })

    def exception(self, err):
        self.publish(topic=self.topicLogger, payload={
            "type":  "exception",
            "line": currentframe().f_back.f_lineno,
            "payload": err
        })

    def publish(self, topic, data):
        self.publish(topic=topic, payload=data)

    def updateThingShadow(self, data):
        self.updateThingShadow(thing=self.thing, payload=data)

    def getThingShadow(self):
        return self.getThingShadow()

File: source/solutions/mini-connected-factory/lambdas/mini-connected-factory-node/ggiot.js
const os = require('os');

class GGIoT {

    constructor(thingName = 'default', prefix = 'mtm') {
        this.thingName = thingName;
        this.prefix = prefix;
        this.topicPrefix = this.prefix + '/' + this.thingName + '/';
        this.topicLogger = this.topicPrefix + 'logger';

        switch (os.platform()) {
            case 'darwin':
                this.iotData = {
                    getThingShadow: (params = {
                        thingName: self.thingName
                    }, callback = (err, data) => {
                        console.log('Default callback');
                    }) => {
                        callback(null, {});
                    },
                    publish: (params = {
                        topic: 'default/topic',
                        payload: JSON.stringify({})
                    }, callback = (err, data) => {
                        console.log('Default callback');
                    }) => {
                        // console.log('iotData.publish: ', params.topic, params.payload);
                        callback(null, 'success');
                    },
                    updateThingShadow: (params = {}, callback = (err, data) => {
                        console.log('Default callback');
                    }) => {
                        callback(null, 'success');
                    }
                };
                break;
            default:
                console.log('GGIOT: not running on a mac');
                this.ggSDK = require('./aws-greengrass-core-sdk');
                this.iotData = new this.ggSDK.IotData();
                break;
        }

        console.log('GGIOT: init for: ' + thingName);
    }

    getThingShadow(callback, thingName = this.thingName) {
        this.iotData.getThingShadow({
            thingName: thingName
        }, (err, data) => {
            if (err) {
                callback(err, data);
            } else {
                callback(err, JSON.parse(data.Payload));
            }
        });
    }

    publish(topic, payload, callback) {
        this.iotData.publish({
            topic: topic,
            payload: JSON.stringify(payload)
        }, callback);
    }

    updateThingShadow(payload, callback, thingName = self.thingName) {
        this.iotData.updateThingShadow({
            thingName: thingName,
            payload: JSON.stringify(payload)
        }, callback);
    }

    info(message, callback) {
        this.publish(this.topicLogger, {
            type: 'info',
            payload: message
        }, callback);
    }

    exception(ex, callback) {
        this.publish(this.topicLogger, {
            type: 'exception',
            payload: ex
        }, callback);
    }
}

module.exports = GGIoT;
